### ISSUE TITLE
Pretraining Large Language Models with NVFP4

### ISSUE THREAD

**Issue by Hiroki11x** [Hiroki11x/Papers#444]
## Links
https://www.arxiv.org/abs/2509.25149

## Authors

<img width="714" height="465" alt="Image" src="https://github.com/user-attachments/assets/e1a84ef4-5dc0-4f24-8b22-9084656be7ff" />

### 🧩3行要約
NVIDIAの論文「Pretraining Large Language Models with NVFP4」は、**4ビット浮動小数点形式（NVFP4）による大規模言語モデル（LLM）の事前学習を安定かつ高精度に実現する手法**を提案したものである。  
従来のFP8訓練と同等の性能を維持しつつ、**計算速度とメモリ効率を大幅に改善**することに成功した。  
特に、**ランダム・ハダマード変換（RHT）、2次元スケーリング、確率的丸め、選択的高精度レイヤー**といった技術の組み合わせが鍵である。

---

<img width="816" height="742" alt="Image" src="https://github.com/user-attachments/assets/58f837d5-a4c8-40f1-b070-8a1bf81e39ca" />

### 1. 研究背景と課題
LLMの訓練コストは現在**数十〜数百ヨタフロップス**規模に達しており、計算効率の改善が喫緊の課題となっている。  
既にFP8精度訓練は普及しているが、**より低精度のFP4（4ビット）**への移行には、収束性や安定性の課題があった。

### 2. NVFP4形式の特徴（Section 2）
従来のMXFP4（microscaling FP4）を改良したNVFP4は、
- **ブロックサイズを32→16要素に縮小**
- **E4M3形式のスケールファクタを導入**
- **FP32スケールをテンソルレベルで併用**
という二段階スケーリング構造を採用し、**動的範囲と精度を両立**。  
これにより、小値のゼロ化を抑えつつ、外れ値の表現精度を改善した（図1参照）。

> **Figure 1（p.3）**  
> 16×32行列をNVFP4形式で格納した例を示す。各16要素ブロックがFP8スケールを共有し、ブロック内の最大値（緑）がFP4最大値に正規化される。

さらに、**NVIDIA Blackwell Tensor Cores**がこの形式をハードウェアレベルでサポートし、FP4計算でFP8比**2〜3倍の演算スループット**を実現する（表1参照）。

---

#### ⚙ モデル設定
- **モデル**：12BパラメータのハイブリッドMamba-Transformer（Nemotron-H系列）  
- **データ**：10兆トークン（最大規模の4-bit訓練）  
- **比較対象**：FP8事前学習モデル

#### 📈 主な結果
- **損失関数**：NVFP4はFP8とほぼ一致し、全学習期間で誤差1〜1.5%以内（図2）

<img width="745" height="385" alt="Image" src="https://github.com/user-attachments/assets/1887f2da-8edb-4ac6-b883-e0dea5569b1e" />

- **下流タスク精度**：  
  MMLU-Pro（5-shot）で**62.58% vs FP8の62.62%**と同等（表2）  
  数学・推論・多言語タスクでもFP8とほぼ同等  

<img width="817" height="284" alt="Image" src="https://github.com/user-attachments/assets/99549226-38c8-4454-ac02-4ff861606c28" />

- **例外**：コーディングタスクでやや劣るが、ノイズ要因と考えられる（図3）

<img width="731" height="457" alt="Image" src="https://github.com/user-attachments/assets/971a9fca-d0ea-4a5a-a5d6-cb7055fc639f" />

> **Figure 2 & 3（p.4–5）**  
> Validation lossと下流タスク精度の推移を示し、FP8とNVFP4がほぼ重なって推移する。

---

### 4. 提案手法（Section 4）
FP4訓練の安定化のため、以下の4技術を組み合わせた。

| 技術 | 概要 | 目的 |
|------|------|------|
| **① 選択的高精度層** | 最後の数層をBF16で保持 | 収束安定化 |
| **② Random Hadamard Transform (RHT)** | 重み勾配入力に適用 | 外れ値分散・分布正規化 |
| **③ 2Dスケーリング** | 16×16ブロック単位で量子化を整合 | 順伝播と逆伝播の整合性維持 |
| **④ 確率的丸め（Stochastic Rounding）** | 勾配量子化に確率的手法 | バイアス低減・収束促進 |

> **Figure 4（p.6）**  
> 各要素を除去したアブレーション研究。4技術のいずれかを削除すると損失が悪化することを示す。

<img width="802" height="462" alt="Image" src="https://github.com/user-attachments/assets/e68b0329-0624-46eb-917e-c0da0c8a0306" />

さらに、**Figure 5（p.7）**では、NVFP4量子化された線形層の計算フローを図示。  
各GEMM演算がNVFP4テンソル入力を用い、勾配計算に確率的丸めを適用している。

<img width="800" height="359" alt="Image" src="https://github.com/user-attachments/assets/c82a4279-2846-4c9a-856b-01a8137413d0" />

---

### 5. 比較実験：MXFP4との性能差（Section 5）
同一条件で8Bモデルを1兆トークン学習した結果：
- **NVFP4の損失誤差：1.5%**
- **MXFP4の損失誤差：2.5%**
MXFP4はNVFP4の精度に到達するために**36%多いトークン（1.36T）**が必要（図6）。  
→ NVFP4は**より高速かつ効率的な学習を実現**。

<img width="792" height="355" alt="Image" src="https://github.com/user-attachments/assets/be6f8a8d-ec1a-4f6a-a7f0-34b89418f826" />

---

### 6. 結論と貢献（Section 6）
- **世界初のマルチトリリオントークンスケール4-bit学習の成功例**を報告。  
- FP8に匹敵する性能を保ちながら、計算・メモリコストを半減。  
- 今後は全線形層のFP4化、Attention・通信経路への拡張、Mixture-of-Experts等での適用が計画されている。

---

### 7. 本研究の新規性と既存研究との差異
- 既存研究（MXFP4, Quartet, FP4-All-The-Way 等）は4-bit訓練の安定性課題を残していた。  
- 本研究は**ハードウェア設計（Blackwell Tensor Core対応）とアルゴリズム設計を統合**し、FP8同等性能を達成。  
- 特に**2DスケーリングとHadamard変換の整合性処理**により、**勾配伝播の連鎖律破れを回避**している点が革新的。

---

### 8. 主なFigureまとめ
| Figure | 内容 | 意義 |
|--------|------|------|
| **Fig.1** | NVFP4フォーマット構造 | FP8＋FP4の2段スケーリング設計 |
| **Fig.2–3** | Loss・精度推移 | FP8と同等の性能を実証 |
| **Fig.4** | アブレーション | 各技術要素の有効性を検証 |
| **Fig.5** | 計算フロー図 | 混合精度GEMM処理の詳細 |
| **Fig.6** | NVFP4 vs MXFP4 | 学習効率でNVFP4が優位 |

---

### 🔍まとめ
本論文は、**NVFP4による4-bit LLM事前学習の実用化に向けた決定的な技術基盤**を築いた。  
これにより、**次世代フロンティアモデルの訓練効率化**（スループット2–3倍、メモリ半減）が可能になり、  
FP4精度での学習が**FP8に並ぶ性能・安定性を持つことを初めて大規模に実証**した。
